### PR TITLE
Extracts logic from apm commands: extract-functions command

### PR DESCRIPTION
**Pull request by ajsantander**
_Ready for review_

Sub PR of #871

### CODE REVIEW

File: packages/aragon-cli/src/helpers/solidity-extractor.js [hunk 1]
```diff
@@ -84,21 +85,71 @@ const getRoles = declaration => {
...
+    return []
+  }
 
-  return funcDecs
-    .filter(dec => modifiesStateAndIsPublic(dec))
-    .map(dec => ({
-      sig: getSignature(dec),
-      roles: getRoles(dec),
-      notice: getNotice(dec),
+  const stateModifyingFunctionDeclarations = functionDeclarations
```
Comment (macor161) [inline packages/aragon-cli/src/helpers/solidity-extractor.js]: Much better names, I like that! 💯

File: packages/aragon-cli/src/helpers/solidity-extractor.js [hunk 2]
```diff
@@ -84,21 +85,71 @@ const getRoles = declaration => {
...
+  // Extract all role ids from the function descriptors.
+  let roleSet = new Set()
+  functionDescriptors.forEach(({ roles }) => roles.forEach(role => roleSet.add(role)))
+  const roleIds = [...roleSet]
+
+  // Parse role ids into objects.
+  // TODO: Name and parameters are currently not being extracted, and it's probably better to get it from an AST instead of the Solidity code. For now, the properties are merely place holders.
+  return roleIds.map(id => ({
+    id,
+    bytes: '0x' + keccak256(id),
```
Comment (macor161) [inline packages/aragon-cli/src/helpers/solidity-extractor.js]:
```suggestion
    bytes: keccak256(id),
```

File: packages/aragon-cli/src/commands/apm_cmds/extract-functions.js [hunk 3]
```diff
@@ -22,37 +20,18 @@ exports.builder = function(yargs) {
...
-  }
-
-  const filename = path.basename(contract).replace('.sol', '.json')
+exports.handler = async function({ cwd, reporter, contract, output }) {
+  const contractPath = path.resolve(cwd, contract)
+  const filename = path.basename(contractPath).replace('.sol', '.json')
   const outputPath = path.resolve(output, filename)
 
-  await writeJson(outputPath, content, { spaces: '\t' })
+  const progressHandler = (step) => {
```
Comment (macor161) [inline packages/aragon-cli/src/commands/apm_cmds/extract-functions.js]: Is the handler necessary? Is it in case we add more steps in the future?

File: packages/aragon-cli/package.json [hunk 4]
```diff
@@ -89,11 +89,11 @@
   "devDependencies": {
     "@aragon/apps-agent": "^2.0.0",
     "@aragon/apps-finance": "^3.0.0",
+    "@aragon/apps-payroll": "^1.0.0",
```
Comment (dapplion) [inline packages/aragon-cli/package.json]: Avoid unnecessary changes in other sections of the package.json to prevent git conflicts

File: packages/aragon-cli/src/lib/apm/extractContractInfoToFile.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+const extract = require('../../helpers/solidity-extractor')
+const { writeJson } = require('fs-extra')
+
+module.exports = async (contractPath, outputPath, progressHandler) => {
```
Comment (dapplion) [inline packages/aragon-cli/src/lib/apm/extractContractInfoToFile.js]: What's the purpose of having a `progressHandler` that gets called once at the end? I would be cleaner to remove it and just log whatever the front-end needs after this async function has resolved.

File: packages/aragon-cli/src/commands/apm_cmds/extract-functions.js [hunk 6]
```diff
@@ -22,37 +20,18 @@ exports.builder = function(yargs) {
...
-  }
-
-  const filename = path.basename(contract).replace('.sol', '.json')
+exports.handler = async function({ cwd, reporter, contract, output }) {
+  const contractPath = path.resolve(cwd, contract)
+  const filename = path.basename(contractPath).replace('.sol', '.json')
   const outputPath = path.resolve(output, filename)
 
-  await writeJson(outputPath, content, { spaces: '\t' })
+  const progressHandler = step => {
```
Comment (dapplion) [inline packages/aragon-cli/src/commands/apm_cmds/extract-functions.js]: See comment on https://github.com/aragon/aragon-cli/pull/870/files#diff-b43061eb5cc2ed67dbd7a209e622dca3R4

File: packages/aragon-cli/src/helpers/solidity-extractor.js (hunk 2 above)
Comment (dapplion) [reply packages/aragon-cli/src/helpers/solidity-extractor.js]: Maybe we can do this in a safe way such as
```js
const hash = hashFunction(id)
return "0x" + (hash.split("0x")[1] || hash)
```
or
```js
const hash = hashFunction(id)
return hash.startsWith("0x") ? hash : "0x" + hash
```
If it's unclear what return format it does

Comment (macor161) [reply packages/aragon-cli/src/helpers/solidity-extractor.js]: @dapplion Yes I think it could be a good idea to abstract this away in case we switch again to another `keccak256` function that doesn't include `0x`. I will create a separate issue for this since the function is used at a lot of different places other than this PR.

Comment (macor161) [reply packages/aragon-cli/src/helpers/solidity-extractor.js]: See #914

File: packages/aragon-cli/package.json (hunk 4 above)
Comment (macor161) [reply packages/aragon-cli/package.json]: Agree that we should avoid changes unrelated to the PR. I left it there because it will probably conflict with `develop` if I remove it.
